### PR TITLE
esp32: enable BLE synchronous events and BLE pairing/boindng

### PR DESCRIPTION
**Pull request by dpgeorge**
This is currently WIP, but all multitests pass (including GAP pairing/bonding) on esp32 with IDF v4.2.

### CODE REVIEW

**Comment by EddieParis** [comment]
I just gave a try to your branch, I successfully paired the ble_bonding_peripheral.py example with my iPhone.
Thanks a lot, this arrives right on time, I just began to work again on an old project where this was needed.

**Comment by EddieParis** [comment]
Sorry to add on your plate, I just noticed that if an unhandeled exception occurs in the irq there is a crash and the ESP reboots.
I had no time (yet) to see if it is a regression.

**Comment by dpgeorge** [comment]
> I just noticed that if an unhandeled exception occurs in the irq there is a crash and the ESP reboots.

Should be fixed by the latest code in this PR (I force pushed to rebase on the lastest master).

**Comment by smurfix** [comment]
Nice! If this is ready to be merged, could you remove the "WIP:" prefix and rebase to master?

**Comment by dpgeorge** [comment]
> If this is ready to be merged, could you remove the "WIP:" prefix and rebase to master?

It's not ready yet, it needs more testing and will not make it in before the next v1.15 release (which should be in the next few days).

**Comment by dpgeorge** [comment]
This is still not 100% reliable, it can lock up the ESP32 (tested with IDF v4.2.2).

**Comment by smurfix** [comment]
Are those lock-ups random? Specifically I just want to do a scan, to tell my home whether it's me at the front door.

**Comment by dpgeorge** [comment]
> Are those lock-ups random?

Yes.  Running the BLE multitests in a loop, it will eventually lock the ESP32 so it does not respond to anything but a hard reset.

**Review by chriscmorgan — COMMENTED** [review]
I am running into this error on the esp32-c3 board when it tries to write to a notify characteristic and its rebooting.

Error:
`GATT` procedure initiated: notify; att_handle=21
assertion "pxQueue->uxItemSize == 0" failed: file "IDF/components/freertos/queue.c", line 1451, function: xQueueSemaphoreTake`

I do not get this error on the esp32 board but that one is having some trouble later in my code trying to connect to the wifi network.

[full error.txt](https://github.com/micropython/micropython/files/7232870/full.error.txt)

**Comment by sahilrastogi94** [comment]
is this still being actively looked into?

File: extmod/modbluetooth.c [hunk 1]
```diff
@@ -1161,6 +1162,72 @@ STATIC mp_obj_t invoke_irq_handler(uint16_t event,
...
+        ts.mp_pending_exception = MP_OBJ_NULL;
+        mp_locals_set(mp_state_ctx.thread.dict_locals); // set from the outer context
+        mp_globals_set(mp_state_ctx.thread.dict_globals); // set from the outer context
+        MP_THREAD_GIL_ENTER();
+    }
+
+    mp_obj_t result = mp_const_none;
+    nlr_buf_t nlr;
+    if (nlr_push(&nlr) == 0) {
+        result = invoke_irq_handler_run(event, numeric, n_unsigned, n_signed, addr, uuid, data, data_len, n_data);
```
Comment (dpgeorge) [inline extmod/modbluetooth.c:1201]: might need to add `mp_sched_lock()` and `mp_sched_unlock()` around this call, to make sure the VM does not go off and handle other tasks

Comment (dpgeorge) [reply extmod/modbluetooth.c:1226]: done

**Comment by dpgeorge** [comment]
I've run extensive tests with this PR using ESP IDF v4.4.1and it works well.

**Comment by sahilrastogi94** [comment]
@dpgeorge can you share an example code to test ble pairing/bonding? I tried the example code, but it seems like the bonding information is not getting saved in nvs. I created a _FLAG_READ_ENCRYPTED characteristic and I have to repair using passkey after every reboot of esp32.

**Comment by FlantasticDan** [comment]
I suspect this pull request removes the need for the following note in the **Pairing and bonding** section of the [bluetooth docs](https://github.com/micropython/micropython/blob/master/docs/library/bluetooth.rst#L721):
```
Note: This is currently only supported when using the NimBLE stack on STM32 and Unix (not ESP32).
```

@dpgeorge if it does I'm happy to open a PR to amend the docs, I'm just awaiting hardware to test for myself and don't want to jump the gun.

**Pull request merged**